### PR TITLE
[Place] Independent Placement Verification

### PR DESCRIPTION
**Pull request by AlexandreSinger**
Created a method that would independently verify the placement in the VPR flow. If a placement passes this verification, it is assumed that it can be used in routing without issue.

By design, this method does not use any global variables (everything needs to be passed in) and recomputes everything that it does not assume. This allows it to be independent of the placement flow being used, so this method can be used in the AP flow as well.

### CODE REVIEW

File: vpr/src/analytical_place/full_legalizer.cpp [hunk 1]
```diff
@@ -393,5 +393,16 @@ void FullLegalizer::legalize(const PartialPlacement& p_placement) {
...
     // Place the clusters based on where the atoms want to be placed.
     place_clusters(clb_nlist, p_placement);
+
+    // Verify that the placement created by the full legalizer is valid.
+    unsigned num_placement_errors = verify_placement(g_vpr_ctx);
+    if (num_placement_errors == 0) {
+        VTR_LOG("Completed placement consistency check successfully.\n");
+    } else {
+        VPR_ERROR(VPR_ERROR_AP,
+                  "Completed placement consistency check, %d errors found.\n"
```
Comment (AlexandreSinger) [inline vpr/src/analytical_place/full_legalizer.cpp:403]: FIXME: This should be %u not %d

File: vpr/src/place/verify_placement.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+ *  @param pl_macros            The placement macros to verify over.
+ *
+ *  @return The number of errors in the macro placement.
+ */
+static unsigned check_macro_placement_consistency(const BlkLocRegistry& blk_loc_registry,
+                                                  const std::vector<t_pl_macro>& pl_macros) {
+    const auto& block_locs = blk_loc_registry.block_locs();
+    const auto& grid_blocks = blk_loc_registry.grid_blocks();
+
+    unsigned error = 0;
```
Comment (AlexandreSinger) [inline vpr/src/place/verify_placement.cpp:149]: FIXME: To be consistent with the rest of the file, this should be `num_errors`.

**Comment by AlexandreSinger** [comment]
@vaughnbetz This is the `check_place` changes we discussed on Friday. I split out all of the invariants from the check_place method and used it to create this method. The idea of the `verify_placement` method is to be an independent check to ensure that all the necessary data structures of placement are filled correctly such that the rest of the VPR flow has no issues (nothing more, nothing less). Please review.

I do have a FIXME in this code which I am not sure about. Are the macros required for the rest of the VPR flow? My intention for this code is that this MUST pass in order for the rest of the VPR flow to be used without issue; but I do not see uses of the pl_macros anywhere outside of placement (and drawing). If they are not used then I think this check should happen somewhere else. What do you think?

**Comment by vaughnbetz** [comment]
If the macros aren't respected generally routing will fail, as the main reason macros are created is for fixed routing patterns (no programmability).  So I'd check them.

File: vpr/src/place/verify_placement.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        if (blk_pr.empty())
+            continue;
+        // Check if the block is placed in its constrained partition region.
+        const t_pl_loc& blk_loc = block_locs[blk_id].loc;
+        if (!blk_pr.is_loc_in_part_reg(blk_loc)) {
+            VTR_LOG_ERROR(
+                "Block %zu is not in correct floorplanning region.\n",
+                size_t(blk_id));
+            num_errors++;
+        }
```
Comment (soheilshahrouz) [inline vpr/src/place/verify_placement.cpp:214]: This is kind of my personal preference, so feel free to ignore it.

I think adding a few empty lines make the code easier to read. For example, I think of checking emptiness of `blk_pr` as a logical step. If the if statement and its comment are preceded and followed by empty lines, I can visually recongnize it as a separte step.

**Comment by soheilshahrouz** [comment]
A few checks are done in `sync_grid_to_blocks()` function. I feel those check belong to `verify_placement.cpp`.

**Comment by vaughnbetz** [comment]
This looks good to me. @AlexandreSinger feel free to merge when you are ready. I tend to agree with @soheilshahrouz that a few blank lines to break up concepts are helpful to readability, but will leave it to you to decide if /where to add any blank lines.

**Comment by vaughnbetz** [comment]
As we discussed the sync_grid_to_blocks() function call could/should be moved inside load_placement, which will make the overall control a bit cleaner. You could also move the post_place_sync () function call into *both* the end of placement and the end of load_placement if you like, as that would again simplify the overall control.

Merge whenever you see fit ...

**Comment by danahow** [comment]
If you want to mark each and every grid covered by a tile, fill in non-root grids with a pointer to the associated root grid. These extra pointers never need to change.

**Comment by vaughnbetz** [comment]
That's what happens now; the grid has non-root node locations point back to the root node.

**Pull request merged**